pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go

```diff
@@ -0,0 +1,436 @@
+	if len(observations) < 2*f+1 {
+		return nil, fmt.Errorf("not enough observations, have %d want %d", len(observations), 2*f+1)
+	}
+
+	currentState, err := a.initializeCurrentState(lggr, previousOutcome)
+	if err != nil {
+		return nil, err
+	}
+
+	report := map[string]any{}
+	shouldReport := false
+
+	for _, field := range a.config.Fields {
+		vals := a.extractValues(lggr, observations, field.InputKey)
+
+		// only proceed if every field has reached the minimum number of observations
+		if len(vals) < 2*f+1 {
+			return nil, fmt.Errorf("not enough observations provided %s, have %d want %d", field.InputKey, len(vals), 2*f+1)
+		}
+
+		singleValue, err := reduce(field.Method, vals)
+		if err != nil {
+			return nil, fmt.Errorf("unable to reduce on method %s, err: %s", field.Method, err.Error())
+		}
+
+		if field.DeviationType != "none" {
+			oldValue := (*currentState)[field.InputKey]
+			currDeviation, err := deviation(field.DeviationType, oldValue, singleValue)
+			if oldValue != nil && err != nil {
+				return nil, fmt.Errorf("unable to determine deviation %s", err.Error())
+			}
+			if oldValue == nil || currDeviation.GreaterThan(field.Deviation) {
+				shouldReport = true
+			}
+		}
+
+		(*currentState)[field.InputKey] = singleValue
+		if len(field.OutputKey) > 0 {
+			report[field.OutputKey] = singleValue
+		} else {
+			report[field.InputKey] = singleValue
+		}
+	}
+
+	// If no AggregationField defines deviation, always report
+	hasNoDeviation := true
+	for _, field := range a.config.Fields {
+		if field.DeviationType != "none" {
+			hasNoDeviation = false
+			break
+		}
+	}
+	if hasNoDeviation {
+		shouldReport = true
+	}
+
+	stateValuesMap, err := values.WrapMap(currentState)
+	if err != nil {
+		return nil, err
+	}
+	stateBytes, err := proto.Marshal(values.ProtoMap(stateValuesMap))
+	if err != nil {
+		return nil, err
+	}
+
+	toWrap, err := formatReport(report, a.config.ReportFormat)
+	if err != nil {
+		return nil, err
+	}
+	reportValuesMap, err := values.NewMap(map[string]any{
+		a.config.OutputFieldName: toWrap,
+	})
+	if err != nil {
+		return nil, err
+	}
+	reportProtoMap := values.Proto(reportValuesMap).GetMapValue()
+
+	lggr.Debugw("Aggregation complete", "shouldReport", shouldReport)
+
+	return &types.AggregationOutcome{
+		EncodableOutcome: reportProtoMap,
+		Metadata:         stateBytes,
+		ShouldReport:     shouldReport,
+	}, nil
+}
+
+func (a *reduceAggregator) initializeCurrentState(lggr logger.Logger, previousOutcome *types.AggregationOutcome) (*map[string]values.Value, error) {
+	currentState := map[string]values.Value{}
+
+	if previousOutcome != nil {
+		pb := &pb.Map{}
+		proto.Unmarshal(previousOutcome.Metadata, pb)
+		mv, err := values.FromMapValueProto(pb)
+		if err != nil {
+			return nil, err
+		}
+		err = mv.UnwrapTo(currentState)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	zeroValue, _ := values.Wrap(decimal.NewFromInt(0))
```

Comment:
Hmmmm... I don't think the result is the same here if the function errors vs if it doesn't, so I would still check the error, or better yet, manually construct the value:

```
var val values.Value
val = values.NewDecimal(decimal.NewFromInt(0))
```